internal/bus/message_pipe.go

```diff
@@ -83,36 +78,76 @@ func (p *MessagePipe) Register(size int, plugins []Plugin) error {
 
 	slog.Info("Finished registering plugins", "plugins", pluginsRegistered)
 
-	p.mu.Unlock()
-
 	return nil
 }
 
-func (p *MessagePipe) DeRegister(pluginNames []string) error {
+func (p *MessagePipe) DeRegister(ctx context.Context, pluginNames []string) error {
 	p.mu.Lock()
+	defer p.mu.Unlock()
 
 	var plugins []Plugin
 	plugins = p.findPlugins(pluginNames, plugins)
 
 	for _, plugin := range plugins {
 		index := getIndex(plugin.Info().Name, p.plugins)
 
-		err := p.unsubscribePlugin(index, plugin)
+		err := p.unsubscribePlugin(ctx, index, plugin)
 		if err != nil {
 			return err
 		}
 	}
 
-	p.mu.Unlock()
-
 	return nil
 }
 
-func (p *MessagePipe) unsubscribePlugin(index int, plugin Plugin) error {
+func (p *MessagePipe) Process(messages ...*Message) {
+	for _, m := range messages {
+		p.messageChannel <- m
+	}
+}
+
+func (p *MessagePipe) Run(ctx context.Context) {
+	p.initPlugins(ctx)
+
+	for {
+		select {
+		case <-ctx.Done():
+			for _, r := range p.plugins {
+				r.Close(ctx)
+			}
+
+			close(p.messageChannel)
+
+			return
+		case m := <-p.messageChannel:
+			p.mu.Lock()
+			p.bus.Publish(m.Topic, ctx, m)
+			p.mu.Unlock()
```

Comment:
I know that this isn't new code but I'm curious why we are locking for the duration of the publish?

The `Publish()` implementation already uses a lock:
```go
func (b *messageBus) Publish(topic string, args ...interface{}) {
	rArgs := buildHandlerArgs(args)

	b.mtx.RLock()
	defer b.mtx.RUnlock()

	if hs, ok := b.handlers[topic]; ok {
		for _, h := range hs {
			h.queue <- rArgs
		}
	}
}
```

Reply:
Good question. I will have another PR going up after this one where I am going to review our use of `sync.Mutex`. I'll address it then